src/Compression/CompressionCodecGCD.cpp

```diff
@@ -0,0 +1,273 @@
+#include <Compression/ICompressionCodec.h>
+#include <Compression/CompressionInfo.h>
+#include <Compression/CompressionFactory.h>
+#include <base/unaligned.h>
+#include <Parsers/IAST.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <IO/WriteHelpers.h>
+#include "Common/Exception.h"
+#include "base/Decimal_fwd.h"
+#include "base/types.h"
+#include "config.h"
+
+#include <cstdint>
+#include <numeric>
+
+namespace DB
+{
+
+template <typename T>
+T gcd_func(T a, T b)
+{
+    while (b != 0)
+    {
+        T c = a % b;
+        a = b;
+        b = c;
+    }
+    if (a == T(-1))
```

Comment:
* prob we need to handle here all negative numbers, not just -1
* if `T` is unsigned this check doesn't make sense. see what happens in `gcd_func<uint8_t>(255, 255)`
* there is `std::gcd` in stl, maybe we could just use it

Reply:
There we can't use std::gcd because it doesn't work with Integers larger than 8 bytes

Reply:
Check boost::integer::gcd